### PR TITLE
add hellaswag and mmlu PyTorch reference eval

### PR DESCRIPTION
**Pull request by karpathy**
it looks pretty bad honestly, the examples themselves, but i'm told it is predictive... hmm
small few examples

```
Context:
 A person is playing bagpipes out in a park. A man and two boys watching from a far. the camera
Endings:
0 (loss: 2.5237) shoots areas all around the park while the bagpipes play.
1 (loss: 4.7942) is covered to the boy that is playing there.
2 (loss: 4.6865) then goes black and we see a black opening scene.
3 (loss: 2.4973) watching the boy and boy as they play a game of drums.
predicted: 3, actual: 0

Context:
 He then takes the knife and sharpens it against the wood piece. He then notices a little black and green poisonous frog sitting next to him. the frog
Endings:
0 (loss: 3.2555) eventually gets to get up and marches out of the room waving its body about.
1 (loss: 3.0197) then starts to get up from the chair and begins walking towards the camera.
2 (loss: 4.5130) then escapes and jumps away.
3 (loss: 3.8115) is acting crazy and attacking him while he is cooking.
predicted: 2, actual: 2

Context:
 A person is seen throwing plaster onto a wall while the camera follows the person close behind. the person
Endings:
0 (loss: 73.8425) continues throwing ingredients down and leads into a person kicking a dirty toilet.
1 (loss: 49.9995) gradually throws more and more onto the wall and the camera pans over the finished product.
2 (loss: 69.1077) then uses paper to spread it all over the wall followed by others putting paper through with the plaster.
3 (loss: 63.4829) is now seen putting wax all along the wall while still looking back to the camera.
predicted: 1, actual: 1

Context:
 We see a bottle of face wash. we
Endings:
0 (loss: 25.8581) see a newspaper story.
1 (loss: 36.2495) see four cookies in a pan.
2 (loss: 36.0045) see a person holding face wash then putting it on their face.
3 (loss: 28.9292) see a tool on a table.
predicted: 0, actual: 2
```

### CODE REVIEW

**Comment by karpathy** [comment]
Note that the typical style of this eval is multiple choice, i.e. present the 4 candidates and have the model pick the best one among them. So the model gets to reason discriminatively because it sees all 4 options.

Here, because these models are so low capability, they cannot do multiple choice. So instead we just create 4 parallel examples and evaluate the likelihood of each individual choice all by itself, and pick the one with the highest likelihood.

This upper bounds your performance well below the peak performance (by 10-20%) but I am told it is still predictive and smooth.

**Comment by karpathy** [comment]
I added MMLU as well:

gpt2 (124M) ~= 25% (chance)
gpt2-xl (1558M) = 27.00% ...

And HellaSwag:

gpt2 (124M) = 28.2%
gpt2-xl (1558M) = 39.22%

TLDR MMLU is way too hard for this scale, but HellaSwag shows some signal, so it is a likely candidate for our GPT-2 repro eval.

**Comment by karpathy** [comment]
Fixed a bug where the loss was off by 1. We have:

**HellaSwag:**

gpt2 (124M)
- eleuther harness reports acc 28.92%, acc_norm 31.14% (multiple choice style)
- this script: 10042 acc: 0.2859 acc_norm: 0.2955 (completion style)

gpt2-xl (1558M)
- eleuther harness reports acc 40.04%, acc_norm 50.89% (multiple choice style)
- this script: 10042 acc: 0.3842 acc_norm: 0.4893 (completion style)

i..e. the dynamic range is ~ok, in that we have:

- 25% random chance
- 124M: 29.55%
- 1558M: 48.93%

and then GPT-3 also reports on Hellaswag and cites similar balpark numbers, though a bit improved, likely due to the much longer training (100B -> 300B tokens) and better dataset. And also because I think it is doing multiple choice instead of simple completion. GPT-3 paper reports 33.7 for ~124M, and 54.7 for ~1558M size. In Appendix H.

**MMLU**

MMLU in comparison is:

- 25% random chance
- 124M: 27.2%
- 1558M: 30.35%

The dynamic range here is a bit too low.

**Comment by karpathy** [comment]
Example eleuther harness eval for gpt2-xl command:

```
lm_eval --model hf     --model_args pretrained=gpt2-xl     --tasks hellaswag     --device cuda:0     --batch_size 16
```
